### DISCUSSION TITLE
Why Amiberry run at 30fps when launcing it from terminal (using distro with no WM, just terminal)

### DISCUSSION THREAD

**Discussion by theshinyknight** [BlitterStudio/amiberry#1131] [Q&A]
Not sure why but I can't get this to work at all.

I am running Armbian and the settings for the display are 1920x1080p60; so it should run at 60 or 50FPS no problems.

But when I run Armbian, the status bar show I am running at locked 30 FPS. Audio is clearly slow down, you can tell that by just listening at how the games run.

Tried to look for a solution but found nothing relevant... I have no way to change resolution as the terminal is running at 1080p@60 already; but when the interface for Amiberry start, it does switch to another resolution (looks like 4:3, as the screen is not completely used); and I can see that the screen has tearing just by moving the mouse. I found no options to change the settings of the UI resolution itself, as changing the video settings does not seem to have any effect at all (tried in window mode and full screen, at different resolutions and nothing seems to solve the problem).

What could the cause be? Is there a way for me to force SDL2 to render Amiberry at a specific resolution? Thanks

**Comment by solskogen** [comment]
Does it run at 60 in X or Wayland? If yes, it's because of SDL. I don't know which version of SDL is shipped with Armbian, but IIRC everything before 2.0.14 has this issue.

**Comment by theshinyknight** [comment]
Hi, I do not have X installed on the system; I just installed the base requirements for running Amiberry, so I would say that it is mainly due to SDL.
Should I get a different SDL version? Should I build SDL from source? How can I change SDL version? Thanks!

**Comment by solskogen** [comment]
How you do that is really up to you, really. I'd rather recommend using X.

**Comment by theshinyknight** [comment]
the problem is that I am running just the terminal environment; isn't Amiberry supposed to run independently from the X server? I have installed the latest SDL version that came with Armbian; I will try to get the latest source and build it too; but that seems to be already newer than 2.0.14 I believe, right?

**Reply by midwan** [reply]
As @solskogen mentioned, Amiberry can run from the terminal but it's SDL2 that has to handle the display. If it has no hardware acceleration, it will still work with software rendering, but it will be much slower - something like what you're describing, usually unable to hit the 50 FPS it requires for PAL emulation.

**Reply by theshinyknight** [reply]
So you think that the 30 fps I see are a result of rendering using software and going slower, and not settings locking the rendering at 30 fps? The screen is an Asus 1080p 75HZ max; made for gaming, but it seems that it is locking the frequency as per signal sent by the pi, more than by lack of power; as the FPS counter does not move at all from the 30, not even an inch no matter what, using just plain WB.

**Reply by midwan** [reply]
That's one possibility, yes. Another would be if you're running a 4k resolution, which the RPI restricts to 30Hz by default. In that case, you can try forcing a 1080p 50Hz resolution, instead.

**Reply by solskogen** [reply]
I'm 99% sure that the reason this happens it due to old SDL.

**Reply by theshinyknight** [reply]
`That's one possibility, yes. Another would be if you're running a 4k resolution, which the RPI restricts to 30Hz by default. In that case, you can try forcing a 1080p 50Hz resolution, instead.`

So I checked the resolution of the monitor, and it is running at 1080p@60; so the app is running at exactly half the frequency of the monitor. I do not have a way to change the frequency of the monitor manually to any other resolution; it is stuck at 1080p@60 currently :( 

One thing that I noticed is that the only parameter I can change and that has effect on the framerate is the refresh parameter in the video settings of Amiberry: by default it is turned off so it is set to 1; if I enable that, it is set to 2 and the framerate goes from 30 to 20; if I set it to 3 it goes to 17 and so on. So that is the only thing affecting the refresh rate, but it reduce it only if I increase the value.

I installed Xorg and xfce too; to see if this fix things. When running Amiberry from the desktop I can see it going above 30; so that seems to solve the problem; but the board is not able to hit 50 at all when running the emulated system at that point, and fluctuate between 40 and 50. I suspect that is due to the fact that now there is a WM running in the back on top of the app in the front; which is why I was trying to not run the WM in first instance, to save resources. Tried to run the minimal config for an A1200 with 14 MHz and 8 MB + 2 MB chip and HDD and even with that I can't keep it stable at 50, even using 640x480 as resolution in full screen.

So basically if I run without WM it is locked at 30, and if I run it with WM it get affected by the WM... it is a no-win scenario I guess

`I'm 99% sure that the reason this happens it due to old SDL.`

I checked that and the version of SDL installed, and which I compiled and linked Amiberry against, is the latest that armbian allow me to install, while running Bookworm , which is 2.26.5. Is this the latest one I can use?

**Reply by solskogen** [reply]
2.26.5 is quite recent and should suffice. What kind of a board are you using?

**Reply by theshinyknight** [reply]
Sorry, gave it a rest for few days while I tried to rebuild, but didn't change anything.

The board is a Banana Pi Zero M2; so technically not the greatest thing ever but not the worst either... Should be able to run at least decently a basic system, looking at its specs.

**Comment by solskogen** [comment]
It can run independently  from X, but due to the problem you're facing it's a lot easier to just run X. Installing a newer SDL manually can open a new can of worms.

**Comment by giantclambake** [comment]
> What kind of a board are you using?

+1 ....not seeing these issues on a rpi3A+ (from CLI, no X) ...but then, I'm not trying to run workbench or anything silly, just Amiga games/demos etc etc 

edit: going by other post, Banana Pi M2 zero

**Reply by theshinyknight** [reply]
Correct; had to build from scratch for Banana Pi Zero M2; and I am not running anything silly either... I went for a plain MagicWB 3.1 to test the basics, so that should run without problems I believe

**Reply by giantclambake** [reply]
I don't share your belief =) The closest hardware I have to this is a rpi3A+  ~ would I (have I) tried running Workbench 3.x on it? The answer is 'no' because I don't believe the hardware could do it without problems. What makes me have this opinion?... due to it's performance trying to run the Sounds of Gnome music-demo disk. Could I be wrong?...absolutely ~ especially when I'm one of those hardcore Amigans who would never, ever run an Amiga (emulated or not) in widescreen mode ; I very really only use 4:3/5:4 aspect displays (windowed amiberry display is in equivalence...a bit like running wine with emulated virtual desktop enabled =)

As [solskogen](https://github.com/solskogen) has alluded, you need to detail a test case, so that any others with a Banana Pi Zero M2 can accurately recreate the usage case ~ currently you've provided insufficient details to allow that testing to happen.

For example, you write "I went for a plain MagicWB 3.1" ...this infers you're running an Amiga Workbench setup, and that could a pre installed and configured Workbench setup (there's several out there), the workbench hdf files from AF, or a Workbench instance you've setup yourself. Simply put, none of us here have any idea what's going on...

...what I do notice is that you haven't mentioned whether JIT is enabled or not ---- I just went through this with the rpi4 and the 'Pimiga' Workbench setup ~ does it work with JIT disabled? No...unsuitably slow, worst experience ever =) Works fine with JIT enabled @ 1080p (works twice as good @ 1280x1024 4:3, but if JIT is disabled it's still intolerably slow =) Thing is, is JIT working for you? (seems broken on the rpi3 build)...but going by recent experience, you're going to need it. 

This all said (and because I'm curious), I will test Workbench on the rpi3A+ using the AF10 release workbench 1.35/2.11/3.11 hdf files. This'll be from CLI boot with console at 1280x1024 (HDMI->VGA db15 -> 19" 4:3 LCD)....and I probably expect wb1.3 will be ok, but I'm definitely going to miss JIT somewhere along the way...ie; a ClassicWB setup is likely asking too much without JIT 8-)

**Reply by theshinyknight** [reply]
Well, I was not expecting to have others share beliefs :) As I do not have a collection of RPI I cannot test every single permutation; but I can use my 4 Kg of matter I carry in my skull and use that magical deduction power we all were gifted with, by looking at data out there and try to infer (which is what I did in the end, to build my "belief")

The first version of Pimiga, came out for RPI3; the RPI Zero 2W does have the same core SOC (BCM2710A1), but less memory. The Banana Pi M2 Zero is comparable in performances with the SOC on the RPI Zero 2W (it is a bit slower), and has faster memory speed (DDR3 vs DDR2), so they should be for most part equivalent.

Armed with this assumption I went on and looked at some videos of Zero 2W running Amiberry and it was running fine; on some groups I was told that you can run ClassicWB no problem (hence why I picked it as it is "standardized", so no risk to try to do something that others can't reproduce... You get the package and use it, you get 1:1 what I am using :) Been around for 20+ years in the field to know how to help fellow computer folks to troubleshoot problems, as I ask the same when troubleshooting my problems on my software).

Sorry if it was not clear which version of MWB I am using; I wrote MagicWB but I should have written **Classic WB** , as that is a fully configured package that require just your own roms (3.1 roms from my A1200, Cloanto edition; I think they are 39.106, need to double check).

So to make it like a proper report, the steps to reproduce would be
- use a BPI M2 Zero stock, download Armbian (https://redirect.armbian.com/bananapim2zero/Bookworm_current) 23.8 bookworm kernel 6.1
- do a full update-upgrade cycle so you have the latest packages
- follow the instructions to get the required packages to build Amiberry (dev packages); check out the git repo and build for H3 platform.
- Get ClassicWB AdvSP package and use your own legal roms and your own WB3.1 to complete its installation  using Amiberry on the M2 Zero board

That will put you in the same state I am now. The emulator runs fine when in WB for most part; I can run WB in 30-45 FPS for most part, which is not a big deal. But if you try to launch  a music player or anything like a game; that is basically unplayable. 

I tried with JIT off and on; I believe I mentioned I tried ALL the options, and the only one that make a difference for most part is the **refresh** one; but that is already turned off so it just make things worst when I turn it on. Reducing resolution, making full screen or in a window or any other audio setting does not seem to have much of an impact. 

Of course reducing the speed of the CPU does add few FPS, but even running a base A1200 with no FPU and 4 MB of ram (2 chip 2 fast Z3) seems to not be feasible. 
So it goes back to how comes that this board that is pretty close to a RPI zero 2W, which has the same SOC of a Pi3, cannot run even the basic Amiga1200; when a Pi3 that has just 512 MB more of ram, can run it fine even with a full X server running behind, using Pimiga?

For reference I have a Pi400 running Pimiga 3, and it is running OK for most part; that board seems to be overclocked, and clearly struggle to run the screen mode at 1080p; and I am not even trying to run WB at that resolution, as I am more of a fan of a simple 4:3 in 640x480 or 800x600 at best for my WB, as this is what I have on my real Amiga hardware. Can't really draw much of a comparison because a Pi400 is a way stronger board than a pi3, so it is way more powerful than the M2 Zero... As such it is logic that it will run better, and make no sense to see what affect that board in the end.

If this board cannot run even the stock Amiga 1200 for WHDLoad and WB, that is OK; but then the RPI zero 2W and the Pi3 should have similar problems (less the Pi3 of course, as it has double the ram)

**Reply by giantclambake** [reply]
...<grin>... thanks for the details ~ you can appreciate how writing 'MagicWB' instead of 'ClassicWB' really does seed confusion in the mind of the reader =^) Unfortunately speaking, saying you tried 'all' options is not the same as indicating which options have been tried, but you get that (some folks, even myself, think they've covered all bases when in fact they [I] hadn't...guilty as self-charged =)

I got to the same state you describe before you posted as a matter of fact ~ I had much fun doing the ClassicWB{Lite/Full/AdvSP} ... enough to say I'll be 'close' to the same setup state ; you haven't listed which ClassicWB installer options you disabled/enabled ...ie; enabling fast copper is a very bad idea... 8) It was working reasonable well ; I didn't test much... the included 'state of the art' demo ran caveat breaking in some effects and some small audio interruptions ... hippoplayer had no issues playing mod files. I didn't try any games (I will typically launch a game directly, not start workbench and launch it from there)...

... I didn't try anything intensive ..ie; AmigaAmp to replay a mp3 file would be a simple example ....and this would be due to the fact that JIT seems broken ATM on rpi3 amiberry builds, and I know for fact that AmigaAmp, under the Pimiga workbench shizzle on rpi4B4g, doesn't work there at all unless JIT is enabled ~ I consider that to be anecdotal evidence enough, to avoid being silly and expecting AmigaAmp to work on rpi3 without JIT ...hehe...

I would like to have tried ClassicWB-P96 but without JIT (and enabling RTG was also crashing...might be related), that's a dead end cause at the now on the rpi3 ~ hopefully midwan can get this going again, and I can continue checking things out =)

>If this board cannot run even the stock Amiga 1200 for WHDLoad and WB

This'll be, in essence, the difference between my own distinctions between whdload tasks, and workbench tasks ~ in my mind, there's a clear delimitation between these 2 usage contexts. As I've been on this planet for more than 6 decades, I recall how 'normal' (and annoying) it used to be playing Amiga games..especially when one had an A590 sidecar, the game couldn't HD install, and it had 4 or more disks, and if you were lucky you had more than 1 floppy-drive =) At this time, whdload had yet come into being. I only mention for the perspective of myself 'being used to' kicking straight from floppy, and not using Workbench at all ~ you'd just be doing the floppy disk version of having a game session on a cartridge based console ; all you're using is the platform hardware (and often time kickstart rom)...ie; you don't boot WB at all really in such a scenario.

I treat whdload game archives, as if they were game cartridges of sorts, and it's amiberry's job to emulate the Amiga machine that 'game cartridge' was designed for ; there is no need for workbench. Due to the fact amiberry's handling of whdload archives is very good, with everything configured via the XML file, on x86-64 I just launch these things directly, either via a links2 filetype association (browse web, click on floppy/whdload/cue/iso file, downloads to /tmp and pipes it to amiberry --autoload or whatever, no persistent local storage of download file)...works at CLI or in X.... I also have mc configured to do that same thing, if you hit enter on a whdload.lha file...

Workbench...I use for anything that ~requires~ WB be present....and TBH for the most part that's more than several demos that fit into the group, Algomusic is like this (but only needs rudimentary WB), but I'm sure everyone has their own list of 'workbench only' stuffs =)

I never combine the 2, just like I can't/couldn't do on my real Amigas ...and IMO not that you'd want to on a low end platform like rpi3 and below ...ie; although sysinfo results aren't really always qualitative, on any given Amiga model/ROM combination comparing sysinfo speed tests, the mips are always a little lower with WB running (as opposed to booting sysinfo.adf) ...which is understandable, but what I can't get my head around is why do this from WB, when you may well get better (game) performance, using amiberry's whdload autoload feature 8^)

Just FTR I'm using this same image on rpi3/4 ..  [raspios_oldstable_full_arm64-2023-10-10/2023-05-03-raspios-bullseye-arm64-full.img.xz](https://downloads.raspberrypi.com/raspios_oldstable_full_arm64/images/raspios_oldstable_full_arm64-2023-10-10/2023-05-03-raspios-bullseye-arm64-full.img.xz)

I'll have to suppose in consideration of JIT not crashing for you, the sun8i build target must be different enough to twart any direct comparisons here on the rpi3 =^/

**Comment by giantclambake** [comment]
Have you got the following option set in conf/amiberry.conf?

`use_sdl2_render_thread=yes`

**Comment by midwan** [comment]
@theshinyknight 
Have you tested with with the latest version of Amiberry? Is it still the same for you?
